### ISSUE TITLE
[RFC] Use prop injection to make Cards aware of their width to help display sublinks with minimal configuration

### ISSUE THREAD

**Issue by OllysCoding** [guardian/dotcom-rendering#6435]
TL;DR Front cards currently aren't aware how much space they are taking up in a container, we can use prop injection to solve this, but should carefully consider this pattern to see if it is a good idea in the long term.

## Sublinks

![sublinks](https://user-images.githubusercontent.com/9575458/201361014-8c20a378-0dc2-456b-b711-51ccb969ac2e.jpg)

Sublinks (or 'supporting content'), are links to stories / cards on fronts that sit as part of an existing card. Editors can provide any number of them for any card added to a front in fronts tool, and the platforms are responsible for figuring out when & how to show them.

There are a number of properties to consider when figuring out how to place a fronts cards
1. Should they show at all? This depends on the card, and we currently decide it by either passing 'supportingContent' to the FrontCard or not. 
2. Should they go alongside the image, or below it?
     1. Based on the position of the image (left/right vs top/bottom)
     2. Depending on the size of the image, we may want to switch between alongside or below based on how many sublinks there are
3.  Should they align horizontally or vertically?
     1. Depends on the width of the card
     2. Depends on the position of the image
     3. Depends on the number of sublinks
     4. Always stack vertically for mobile/tablet
4. What's the max number of sublinks it cane show?
     1. Depends on the size of the card
5. Should trail text be kept? (some cards wont show trail text if there is supporting content)

At the moment, the card doesn't know what size it is - this makes it quite difficult for us to determine how to display the sublinks.

## Injecting the percentage prop through to FrontCard

Prop injection allows us to add props to the children of a react component,

```tsx
<LI percentage="50%">
     <FrontCard ... />
</LI>
```

In the above example, `FrontCard` has no way of knowing what % it is taking up, without us passing the percentage prop a second time.

However, we can set up a new function `injectProps`:

```ts
import type { ReactNode } from 'react';
import { Children, cloneElement, isValidElement } from 'react';

export const injectProps = (
	children: ReactNode,
	props: Partial<unknown>,
): ReactNode => {
	return Children.map(children, (child) => {
		if (isValidElement(child)) {
			return cloneElement(child, props);
		}
		return child;
	});
};
```

And then in LI, we can pass the percentage property through to its children: 

```tsx
// LI.tsx
return (
	<li css={[...]}>
		{injectProps(children, { parentPercentage: percentage })}
	</li>
);
```

And now, if Card.tsx accepts an optional prop of `parentPercentage`, it will receive the percentage passed to LI, and be able to use it for decision making.

### Passing through multiple LIs & ULs

It's quite common to have a setup like so:

```tsx
<LI percentage="50%">
     <UL>
            <LI percentage="50%"><FrontCard ... /></LI>
            <LI percentage="50%"><FrontCard ... /></LI>
     </UL>
</LI>
```

In this case it's a bit more difficult, we can tell by reading this that the card will take up 25% of the overall container, but our current solution will pass 50% as the `parentPercentage`

To solve this, we first must expand the UL component to take `parentPercentage` as a prop, and use the `injectProps` to pass that directly through to any children. As UL doesn't take a percentage prop of its own, it's essentially just acting as a pass through of any percentages set on LI components.

Then we must expand the LI component to be capable of taking `parentPercentage` as a prop, and then 

```tsx
// LI.tsx
const getPercentage = () => {
	if (parentPercentage && percentage)
		return getResultingPercentage(parentPercentage, percentage);
	if (parentPercentage && !percentage) return parentPercentage;
        return percentage;
};

return (
	<li css={[...]}>
		{injectProps(children, { parentPercentage: getPercentage() })}
	</li>
);
```

## What does this let us do?

Now that we know the card percentage, we can start to put together functions that help us make rendering decisions about sublinks, some examples of this might be: 

```ts
const decideSublinkCount = (width: CardPercentageType) => {
     switch (width) {
          case "100%":
               return 4
          case "75%":
               return 3
           default:
               return 2 
     }
}

const decideSublinkPosition = ...

const decideSublinkAlignment = ...
```

A nice part of this solution is that it keeps all of the sublink display logic together, in a set of concise functions that _should_ be easy to reason about.

## Can this be used elsewhere?

This could prove to be useful to give use more 'accurate' images sizes in [CardPicture](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/src/web/components/CardPicture.tsx#L21), which can currently only make best guesses off of the image size.
 
## What are the disadvantages?

This would be a new pattern in DCR, and one that's potentially difficult to reason about as a new engineer, as it might be hard to discover how the card knows what its percentage width is.

Another issue is that the percentage width is subjective - if we changed the width of containers for example, we might need to re-reason a lot of the logic.

While we can ensure some type safety, it may be difficult to prevent mistakes, engineers could try and set the 'parentPercentage' prop directly, not realising it's set automatically (perhaps an ESLint rule could prevent this). Additionally, adding extra elements to the UL > LI > FrontCard flow could cause it to break, meaning the pattern may be relatively fickle. 

Finally, it lifts control out of the configuration of the card itself. When you're creating or working on a container, you don't have control over the rules deciding how the sublinks are shown. In some situations this is good, it removes complexity from container creation, and introduces consistencies between containers. However, this also means that if it wasn't working as expected, you'd be unable to get your individual container to look the way you expect, and instead need to investigate the full logic of sublinks.

### Use a supportingContent 'config'

An alternative would be to create some kind of supportingContent 'config' that can be passed to cards, enabling individual configuration of how sublinks display: 

```tsx
<FrontCard
    ...
    supportingContent={{
         trails={card.supportingContent}
         maxTrails: 4 // Maximum number of supported trails
         alignHorizontalCount: 3 // Switches to displaying horizontally when there are this number of trails or more
         inlineUntilCount: 3 // Switches from being inline to to outside the card when this number of trails or more are present
         hideTrailText: true // Hides trail text if there is supporting content
    }}
/>
```

A clear disadvantage of this approach is that it requires anyone creating a card  to have an understanding of what these properties are, and how their card should interact with them all, depending on the number of sublinks provided, and how they interact with all the other card properties (width, image position, etc).

That said, it also makes the rules a lot more clear, rather than abstracting them away.

### A hybrid between config & abstraction

Closer to what we have already, we could achieve a hybrid between config & abstraction, where some hints or pieces of information are given by FrontCard props, and others can be abstracted away. I don't have a clear example of this, but it could end up looking something like;

```tsx
// some card def:
<FrontCard
	trailText={
		// Only show trail text if there is no supportContent
		cards[0].supportingContent === undefined ||
		cards[0].supportingContent.length === 0
			? cards[0].trailText
			: undefined
	}
	supportingContent={cards[0].supportingContent?.slice(0, 4)}
	supportingContentAlignmentHint={
		cards[0].supportingContent &&
		cards[0].supportingContent.length > 3
			? 'horizontal'
			: 'vertical'
	}
/>

// Card.tsx
const decideSublinkPosition = (
	supportingContent?: DCRSupportingContent[],
	imagePosition?: ImagePositionType,
	imageSize?: ImageSizeType,
): 'inner' | 'outer' | 'none' => {
	if (!supportingContent || supportingContent.length === 0) return 'none';
	if (
		imageSize === 'jumbo' &&
		(imagePosition === 'left' || imagePosition === 'right')
	) return supportingContent.length > 3 ? 'outer' : 'inner';
	if (imagePosition === 'top' || supportingContent.length > 2) return 'outer';
	return 'inner';
};
```

Unfortunately, this may offer the 'worst' of both worlds, with the interface for FrontCard remaining confusing and requiring knowledge and understanding of how sublinks work, and abstraction meaning that you still can't get an immediate understanding of how your card will look in practice.

## Feedback please!

There's a lot covered here, and without a doubt this is problem does not have an 'easy' solution, so we really want to consider:
1. What do we want to support into the future?
2. What will be best for engineers to understand & work on in the future?

Please provide your thoughts on the injection pattern, on the sublink alternatives, and if you have any other ideas!

**Comment by ob6160** [comment]
Great work! I like how you gave so many directly applicable examples :)

To kick a discussion off, I was wondering.. has React Context been considered as one of the alternatives to injecting props? 

It seems (at first glance) to be a good fit for this kind of problem. I've seen frameworks like Chakra UI use it to great effect. An example being how they use it to pass down props from their `RadioGroup` into individual `Radio` components without having to prop-drill (ref: https://github.com/chakra-ui/chakra-ui/tree/main/packages/components/radio/src)

Taking `<FrontCard />` needing to be aware of the percentage of its parent as an example, I'm wondering about what the pros/cons of an approach like the following would be _(note: very shortened code examples which may be missing implementation details as I quickly typed them out)_:

1. Setup a React Context provider in the parent list component:

```tsx
interface CardListContext {
	percentage: number;
    // other stuff that you want to "inject"
}

const [CardListProvider, useCardListContext] =
  createContext<CardListContext>({
    name: "CardListContext",
    strict: false,
  })
```

2. Pass that context down using the provider:

```tsx
<CardListProvider data={{percentage: 50}}>
  <li><FrontCard /></li>
  <li><FrontCard /></li>
  <li><FrontCard /></li>
</CardListProvider>
```

3. Access that context from inside the FrontCard component (it will be undefined if it there is no provider above it):

```tsx
const { percentage } = useCardListContext();

// do something with `percentage` if it is defined
```

The immediate drawback to this approach that I can see is that it ties some of the functionality of `FrontCard` to the presence of a `CardListProvider` that sits above it. It strikes me that it might be important to indicate the relationship between the two.. perhaps using a scheme like Chakra do in the example I gave earlier.. `CardGroup`/`FrontCard` to draw a parallel to their `RadioGroup`/`Radio?

Anyway, I just thought I'd throw this idea into the discussion because this struck me as a case where React Context might fit in really well (at least ostensibly 😉)

**Comment by i-hardy** [comment]
A pattern we've adopted in Conversions while building the new supporter prop checkout is making a lot of use of [render props](https://reactjs.org/docs/render-props.html). This is partly because it enables us to write Redux and/or payment provider-related logic separately from UI components that are more subject to A/B testing and design changes, but it also enables us to have a parent component calculate some props for to its children while other props come from elsewhere.

In this case a render prop setup might look something like:

```typescript
// LI.tsx
function LI({ percentage, render }: LIPropsType) {
	// percentage calculation
	return (
		<li css={...}>
			{render(percentage)}
		</li>
	}:
}

// Front.tsx
<LI
	percentage="50%"
	render={(childPercentage) => (
		<FrontCard size={childPercentage} {...otherCardPropsFromElsewhere} />
	)}
/>
```

This can also be a lot more flexible if you need to take that initial `childPercentage` and do something else with it rather than just directly passing it into a component, such as in your UL example. You can also just straight up use `children` as your render prop, and write something like:

```typescript
<LI percentage="50%">
	{(childPercentage) => <FrontCard size={childPercentage} {...otherCardPropsFromElsewhere} />}
</LI>
```

but I'm not personally fond of having `children` be a node in some components and mysteriously sometimes a function in others.

**Comment by OllysCoding** [comment]
> @ob6160 To kick a discussion off, I was wondering.. has React Context been considered as one of the alternatives to injecting props?

I hadn't considered it at all, but am quite intrigued from the example you sent me! 

One thing I'm curious about is handling multiple embedded LIs
```
<UL>
     <LI percentage="50%"><FrontCard .../ ></LI>
     <LI percentage="50%">
          <UL>
               <LI percentage="50%"><FrontCard ../></LI>
               <LI percentage="50%"><FrontCard ../></LI>
          </UL>
     </LI>
<UL>
```

In this case we'd need context to either be able to do the calculation behind the scenes when the context is set twice (do more advanced implementations support this?), or have multiple contexts, which seems like it could be brittle.

---

This aside, I really quite like this solution, as being quite verbose, it does it in way that makes it super clear what happens! Thanks for the suggestion :)

**Comment by OllysCoding** [comment]
@i-hardy Thanks for this information! This was a pattern I hadn't considered either!! 

I wanted to imagine what multiple levels of ULs and LIs might look like, just so I have a good understanding, it looks like it could be a tad confusing, but workable:

```tsx
<LI 
    percentage="50%"
    render={(percentage) => (
          <UL>
               <LI 
                    parentPercentage={percentage}
                    percentage="50%"
                    render={(percentage2) => <FrontCard size={percentage2} /> }
               />
          </UL>
     )}
/>
```
In the above example we assume the second LI is combining the percentages to output percentage2, but we could also call the combine() function in the FrontCard definition as well.

My biggest concern would be having multiple of these render props embedded in one-another.

**Comment by mxdvl** [comment]
Thought: an alternative would be being extremely explicit about a card’s size! If a card takes in the number of columns it spans along with the image position, the need for this pattern vanishes?

---

> A pattern we've adopted in Conversions while building the new supporter prop checkout is making a lot of use of [render props](https://reactjs.org/docs/render-props.html).

This is a very interesting approach, @i-hardy! I think this is easier to reason about than props injection.

**Comment by arelra** [comment]
The solutions here are really clever! If they solve a problem then great but I feel they could lead us a little away from some core [DCR principles](https://github.com/guardian/dotcom-rendering/blob/main/dotcom-rendering/README.md#concepts) wrt to simplicity, prop drilling and avoiding abstractions.

I am not super sure I have grokked this entirely but iiuc the core problem is:

> Given the size of a container how should I render my children

or:

> Given the size of a FrontCard how should I render SubLinks

Is trying to work with the generic `<UL><LI>` pattern increasing the complexity here and leading to having to rewrite the props of children?

Could we instead have named components that encapsulate the presentational structure and hence logic of how many children to render via mapping:

Rough outline:

```js
// FrontCard.tsx
const FrontCard = ({size} ) => {
  // calculate here
  const maxSubLinks = (size) => 
  const sublinkPosition = ...
  const sublinkAlignment = ...  
  return (
    <>
      ...
      <SubLinks subLinks={subLinks.slice(0, maxSubLinks)} />
    <>
  );
}

// SubLinks.tsx
const SubLinks = ({ maxSubLinks, subLinks }) => {
  return (
    <UL>
      {subLinks.map((subLink) =>
        (<LI><SubLink subLink={subLink} /></LI>)
      }
    </UL>
  );
}
```

_If_ that works for the use case you outlined I think it might be more approachable than an involved abstraction.

**Comment by OllysCoding** [comment]
Closing as the work for solving the underlying problem was ultimately completed in https://github.com/guardian/dotcom-rendering/issues/6691
